### PR TITLE
Passing tests

### PR DESCRIPTION
**Pull request by FynnBe**
clean up transformations and loss functions
add tests

requires https://github.com/bioimage-io/python-bioimage-io/pull/7

### CODE REVIEW

File: pybio/torch/transformations/base.py [hunk 1]
```diff
@@ -1,42 +1,39 @@
 # TODO would be nice to have something similar to inferno's batch_function etc.
 # functionality, but more flexible w.r.t implementing different dimensions.
+from typing import List, Optional, Sequence, Tuple
 
+import torch
 
-class IndependentTransformation:
+import pybio.core.transformations
```
Comment (constantinpape) [inline pybio/torch/transformations/base.py]: Do we need a transformation base class in `pybio.core`?
Right now this feels like it's just producing more boilerplate code to me.
**Edit:** It seems like you have moved the implementation of the `__call__` method there, but I am not sure if this is the good idea,
because it's (implicitly) imposing the specific transformation mechanism on everyone who would implement the `pybio` API.

File: pybio/torch/transformations/base.py [hunk 2]
```diff
@@ -1,42 +1,39 @@
...
 
-def apply_transformations(transformations, *tensors):
-    """ Helper function to apply a list of transformations to input tensors.
-    """
-    if not all(callable(trafo) for trafo in transformations):
-        raise ValueError("Expect iterable with callables")
-    for trafo in transformations:
-        tensors = trafo(*tensors)
-    return tensors
+class Loss(pybio.core.transformations.Loss):
```
Comment (constantinpape) [inline pybio/torch/transformations/base.py]: Same as transformation, I don't know if it's a good idea to move the parent class with some implementation to the toplevel library.

File: specs/transformations/BCELoss.transformation.yaml [hunk 3]
```diff
@@ -32,5 +33,5 @@ outputs:
     - name: loss
       data_type: float32
       data_range: [0, 1]
-      # axes: null   we don't care about the axes here, so this should be optional
+      axes: "c"
```
Comment (constantinpape) [inline specs/transformations/BCELoss.transformation.yaml]: I think that the BCELoss will just return a scalar, so having the `axes: "c"` is not really correct.

File: pybio/torch/transformations/augmentations.py [hunk 4]
```diff
@@ -14,5 +14,5 @@ def set_next_state(self):
         self.k = np.random.randint(0, 4)
 
     # FIXME only works for 2d inputs
-    def apply_transformation(self, tensor):
+    def apply_to_one(self, tensor):
```
Comment (constantinpape) [inline pybio/torch/transformations/augmentations.py]: Technically this does not depend on pytorch either, so could go to `pybio`.

File: pybio/torch/transformations/base.py (hunk 1 above)
Comment (FynnBe) [reply pybio/torch/transformations/base.py]: I think it would be great if torchbio's transformations behave the same way pybio ones do, therefore I would inherit from the pybio base transformation. It's there anyway for the framework agnostic transformations.

there should not be any `__call__` methods left for transformations... ?

Comment (FynnBe) [reply pybio/torch/transformations/base.py]: oh, sorry I confused Transformation with BaseTransformation... yeah.. let's get rid of the latter, it seems to cause confusions^^

File: pybio/torch/transformations/base.py (hunk 2 above)
Comment (FynnBe) [reply pybio/torch/transformations/base.py]: Loss is now only in pybio.torch, not in core...

File: specs/transformations/BCELoss.transformation.yaml (hunk 3 above)
Comment (FynnBe) [reply specs/transformations/BCELoss.transformation.yaml]: true, removed axes and changed shape to
```
shape: []
```

File: pybio/torch/transformations/augmentations.py (hunk 4 above)
Comment (FynnBe) [reply pybio/torch/transformations/augmentations.py]: yes, should we move SynchronizedTransformation to `pybio.core` as well then?
And what about assuming the Transformation is independent (remove additional IndpendentTransformation)?

File: pybio/torch/training/simple.py [hunk 5]
```diff
@@ -6,41 +9,32 @@
...
+from pybio.spec.spec_types import ModelSpec
+
 from torch.utils.data import DataLoader
 from pybio.torch.transformations import apply_transformations
 
-
-# TODO config is just a stub object right now, adapt this to
-# the actual config object from pythonbioimageio
-def simple_training(config, n_iterations=500, batch_size=4, num_workers=2, out_file="./weights.pytorch"):
+def simple_training(model_spec: ModelSpec, n_iterations: int, batch_size: int, num_workers: int, out_file: Union[str, Path, IO[bytes]]) -> torch.nn.Module:
```
[the remark this answers is not in the document]
Comment (FynnBe) [reply pybio/torch/training/simple.py]: example in model.yaml
```yaml
...
training:
 ...  
 source: pybio.torch.training.simple.simple_training
  required_kwargs: [model_spec]
  optional_kwargs: {n_iterations: 25, batch_size: 4, num_workers: 2, out_file: ./weights.pytorch}
```

File: manifest.yaml [hunk 6]
```diff
@@ -1,7 +1,8 @@
+models:
+    - specs/models/unet2d/nuclei_broad/UNet2DNucleiBroad.model.yaml
 transformations:
-    - specs/transformations/BCELoss.transformation.yaml
+    - specs/losses/BCELoss.transformation.yaml
```
Comment (constantinpape) [inline manifest.yaml]: I thought we agreed that there is no difference between transformatins and losses, at least on the spec level?

File: pybio/torch/training/simple.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (constantinpape) [reply pybio/torch/training/simple.py]: I just meant breaking the line between arguments:
```py
def simple_training(model_spec: ModelSpec, 
                                n_iterations: int,
                                ...)
```
instead of 
```py
def simple_training(model_spec: ModelSpec, .....):
```
The line is too long right now ;).
(We should set up black liniting at some point to take care of these things, but let's not get ahead of ourselves.)

File: manifest.yaml (hunk 6 above)
Comment (FynnBe) [reply manifest.yaml]: true, but also specs may lie anywhere. For more organization I split the transformations specs into the two folders transformations and losses...

**Comment by FynnBe** [comment]
> Looks good overall.
> Two points:
> 
> * Do we need to have anything separate for losses on the spec level? Right now it follows the same `List[tensors] -> List[tensors]` logic as all other transformations, so I don't see why we need to single losses out in the specs.

only for type checking and organizing the code into smaller, oversee-able modules. However they do direclty inherite from CombinedTransformation.

> * We need an integration test that uses the simple train loop to actually train a model. This can either go into `test_UNet2dNucleiBroad` or go into a separate test, however you prefer it.

Currently the simple training loop is tested (I should better say used) in `tests/test_specs/test_models/test_UNet2dNucleiBroad.py`

File: pybio/torch/training/simple.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FynnBe) [reply pybio/torch/training/simple.py]: in the current commit it's blacked properly:
https://github.com/bioimage-io/pytorch-bioimage-io/blob/aafff158e7bd46c9c95e919b2fc99a8bcd64afcd/pybio/torch/training/simple.py#L20-L22

[the remark this answers is not in the document]
Comment (constantinpape) [reply pybio/torch/training/simple.py]: ok, that's much better then the initial commit already

File: manifest.yaml (hunk 6 above)
Comment (constantinpape) [reply manifest.yaml]: Hmm, I still find this potentially confusing, if we don't have the concept of `loss` in the spec, we shouldn't put it into another folder here.
Sure, it's allowed, but doesn't mean it's good practice.

**Comment by constantinpape** [comment]
> Currently the simple training loop is tested (I should better say used) in tests/test_specs/test_models/test_UNet2dNucleiBroad.py

I see. I was confused because the train function is not explicitly called, but is pushed into some method of the model. A more explicit test for the function would be nice, but I think doesn't need to happen in this PR.
